docs/proposals/20210222-kubelet-authentication.md

```diff
@@ -0,0 +1,321 @@
+
+#### Provider Specification
+
+#### All providers
+- All providers MUST insert a ProviderID within the KubernetesNodeProviderIdentifierOID extension attribute of the CSR
+- All signer names MUST be filled in by the provider’s controller in InfraCluster.Status.KubeletClientCertificateSigner and InfraCluster.Status.KubeletServingCertificateSigner
+
+#### Insecure providers
+- An insecure provider CANNOT implement certificate rotation or kubelet serving certificate signing
+- InfraCluster.Status.KubeletClientCertificateSigner MUST be set to cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure
+- An insecure provider MUST use the cluster.x-k8s.io/kube-apiserver-client-kubelet-insecure signer/v1alpha4
+
+#### Secure providers
+- A secure provider MUST implement certificate rotation and kubelet server certificate signing
+- A provider must register signers of:
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-<provider>/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-<provider>/v1alpha4`
+- A secure provider MUST implement a secure attestation mechanism, based upon PEM-encoded blocks within the Certificate Signing Request.
+- Where a secure provider’s attestation mechanism does not include a nonce or timestamp to protect against replay attacks, the mechanism MUST implement a secondary time-limited attestation (e.g. AWS Instance Identity document + AWS HMACv4 signature).
+- A provider’s signer MUST run on the management cluster
+
+#### TPM based providers
+- A TPM provider MUST use the following certificate signers
+    - `cluster-x-k8s-io/kube-apiserver-client-kubelet-tpm/v1alpha4`
+    - `cluster-x-k8s-io/kubelet-serving-tpm/v1alpha4`
+- A TPM provider MUST annotate new CSRs as follows:
+    - Key: cluster-x-k8s-io/tpm-endorsement-key/v1alpha4
+    - Value: Platform-specific endorsement key (e.g., retrieved from GCP Shielded VM API or VMware vCenter)
+
+#### Kubeadm
+Since this proposal essentially takes over part of the node registration process from kubeadm, we will require the following changes:
+- kubeadm COULD allow opt-out of kubeadm setting up ClusterRoleBindings between the system:nodes group and the `system:certificates.k8s.io:certificatesigningrequests:selfnodeclient` permission, so that certificate renewals must go through re-attestation.
+- Kubeadm COULD allow opt-out of kubeadm setting up the bootstrapper and node groups and bindings.
+
+This is deferred to a future Kubeadm design and release, and for this proposal, we will add fields to KubeadmControlPlane to remove these node groups and bindings post control plane initialisation.
+
+#### Changes to KubeadmControlPlane
+
+```
+
+```
+
+#### Kubelet authenticator flow
+
+The authenticator will be responsible for both updating the kubelet client and serving certificates.
+
+![kubelet auth](images/kubelet-authentication/kubelet-auth.png)
```

Comment:
I think the false edge should not go to `Begin kubelet serviing CSR process`:

![image](https://user-images.githubusercontent.com/4662360/113105944-76768200-9202-11eb-997f-b19b16732bee.png)

Reply:
It's actually correct in the plantuml that i'm about to upload, it just rendered really badly
let me try and sort out.

Reply:
rendered really badly. i  changed "true" , "false" to more specific wording and separated out the two diagrams.